midend/copyStructures.h

```diff
@@ -22,6 +22,21 @@ limitations under the License.
 
 namespace P4 {
 
+struct CopyStructuresConfig {
+    /// Specific targets may allow functions or methods to return structs.
+    /// Such methods will not be converted in this pass. Setting the
+    /// errorOnMethodCall flag will produce an error message if such a
+    ///  method is encountered.
+    bool errorOnMethodCall = true;
+
+    /// Do not only copy normal structures but also perform copy assignments for headers.
+    bool copyHeaders = false;
+    /// Also expand header union assignments.
+    /// TODO: This is only necessary because the copy structure pass does not correctly expand
+    /// header unions for some back ends.
```

Comment:
Can you elaborate a bit what this expansion means?

Reply:
Let me rethink this. The context is https://github.com/p4lang/p4c/pull/4853#issuecomment-2272870543

The way P4Testgen interprets this expansion leads to behavior that is not correct. I do not recall whether other back ends have a similar problem but P4Testgen just interprets the spec which says: 
```
u.hi.setValid(): sets the valid bit for header hi to true and sets the valid bit for all other headers to false, which implies that it is unspecified what value reading any member header of u will return.
```

Reply:
Hmm, given that it is fundamentally broken, we might as well say that `CopyStructures` does not support expansion of header union assignments and error in such case. Still, I'm worried how this would impact backends that use it (if any).

It seems to me the only way to do this in midend would be to do consecutive `if`s` for `isValid()` for all alternatives, which could lead to inefficient code in many backend. But backends can always opt-out of expanding headers.

BTW. how does testgen process header union assignments if it expands assignments of headers, but not of header unions?

---

Also, this PR should probably have a `breaking-change` tag because of the changes in `CopyStructures`.

Reply:
I think I will split the change to CopyStructures out again and mark it breaking, then rebase the P4Testgen fixes on top.

Reply:
>BTW. how does testgen process header union assignments if it expands assignments of headers, but not of header unions?

Header union and header assignments should now be handled by https://github.com/p4lang/p4c/blob/main/backends/p4tools/modules/testgen/core/small_step/cmd_stepper.cpp#L68 and https://github.com/p4lang/p4c/blob/main/backends/p4tools/common/core/abstract_execution_state.cpp#L186C30-L186C46

Iirc CopyStructures is technically not needed by P4Testgen but there are other passes it uses (ParserUnroll?) which require it.

Reply:
> Iirc CopyStructures is technically not needed by P4Testgen anymore but there are other passes it uses (ParserUnroll?) which require it.

It might be worth checking if we can just disable it in testgen (or leave it just for structures).